src/box/box.cc

```diff
@@ -793,6 +798,80 @@ wal_stream_apply_dml_row(struct wal_stream *stream, struct xrow_header *row)
 	return -1;
 }
 
+struct rlist nodes_rows[VCLOCK_MAX];
+
+/*
+ * When restoring the log is read row-by-row. However, it is necessary
+ * to store the rows for correct further recovery. For example, with
+ * mixed transactions. The following function saves the newly arrived
+ * row to the rlist, which is stored in array is nodes_rows in cell is
+ * replica_id-1. As soon as a row arrives with the is_commit flag set,
+ * the corresponding transaction tries to apply. If an error occurs the
+ * transaction is rolled back. And we continue to work with the next
+ * transaction.
+ */
+static int
+wal_stream_apply_mixed_dml_row(struct wal_stream *stream,
```

Comment:
You should also check that once recovery is finished there are no unfinished transactions left in your `nodes_rows` array, like simple recovery does it:
https://github.com/tarantool/tarantool/blob/9fcafbd14b962d9f844a9d114518fa690753f276/src/box/box.cc#L4526-L4533

Reply:
Done. Added test.

Reply:
Done